flink-ml-lib/src/main/java/org/apache/flink/ml/common/datastream/SortPartitionImpl.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.common.datastream;
+
+import org.apache.flink.api.common.state.ListState;
+import org.apache.flink.api.common.state.ListStateDescriptor;
+import org.apache.flink.api.common.typeutils.TypeComparator;
+import org.apache.flink.ml.common.utils.ComparatorAdapter;
+import org.apache.flink.runtime.state.StateInitializationContext;
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.BoundedOneInput;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+
+import org.apache.commons.collections.IteratorUtils;
+
+import java.util.List;
+
+/** Applies sortPartition to a bounded data stream. */
+class SortPartitionImpl {
```

Comment:
Having one dedicated file for a package private static class seems a bit overkill. This is a rare pattern in Flink. The typical approach is to put those classes in the same file as the public static method that uses them (e.g. DataStreamUtil).

It might make sense to have dedicated file if the number of static class/method is more than 4 (e.g. AllReduce).

Maybe it is simpler to move `SortPartitionImpl/DistinctImpl/MapPartitionImpl` to DataStreamUtils.

Reply:
If we only have three utlitity function (i.e., `SortPartitionImpl/DistinctImpl/MapPartitionImpl`), I am fine to move them to `DataStreamUtils`.

The question here is that we clearly will have more utility functions for DataStream in the future. Adding all of them to `DataStreamUtils` may make `DataStreamUtils` to complex.

An alternative is to add a new package (`org.apache.flink.ml.common.datastream.impl`) and put all these implementation under `impl` package. What do you think?

Reply:
The problem you described here is common to most projects (e.g. Flink). Can we follow the existing practice in e.g. Flink instead inventing new patterns?

Is there any util class in popular projects like Flink/Kafka/Spark that explicitly moves the implementation of its static methods to individual `*impl` classes?

Reply:
Sure, let's follow the existing practice.